File: moderate/cycle_detection.py
import fileinput

for line in fileinput.input():
    st = line.split()

    f = []
    j = {}
    while st:
        i = st.pop(0)
        if i not in j:
            if not st:
                break
            j[i] = st[0]
            if f:
                break
        else:
            if i in f:
                break
            f.append(i)
    if f:
        print ' '.join(i for i in f)

File: moderate/cycle_detection.rb
File.open(ARGV[0]).each_line do |line|
  s = line.chomp.split
  while s.length > 1 and s.count(s[0]) == 1
    s.delete_at(0)
  end
  if s.length == 1
    puts
  else
    t = [s.delete_at(0)]
    s.each do |x|
      break if x == t[0]
      t << x
    end
    puts t.join(" ")
  end
end
